### ISSUE TITLE
[lldb] No pretty print for the coroutine frame type.

### ISSUE THREAD

**Issue by hokein** [llvm/llvm-project#69309]
The lldb doesn't seem to know about the coroutine frame type (the structure type [`<function_name>.coro_frame_ty`](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Coroutines/CoroFrame.cpp#L1134) emitting in the clang middle-end). See the poor print results below.

```
$ ./bin/clang++ -std=c++20 -g /tmp/t.cpp -o debugging-example 
$ lldb debugging-example 
...
(lldb) r
...
Process 1790445 stopped
* thread #1, name = 'debugging-examp', stop reason = signal SIGTRAP
    frame #0: 0x00005555555555fc debugging-example`coro_task(v=1) at t.cpp:46:3
   43    task coro_task(int v) {
   44     int a = v;
   45     __builtin_debugtrap();
-> 46     co_return;
   47   }
   48  
   49   int main() {
(lldb) p __coro_frame
(_Z9coro_taski.coro_frame_ty)  {}
```

The gdb gives better results on printing the `__coro_frame`:

```
(gdb) p __coro_frame
$1 = {__resume_fn = 0x555555555bf0 <detail::chain_fn<0>()>, __destroy_fn = 0x555555555d90 <detail::chain_fn<0>()>, __promise = {continuation = {_M_fr_ptr = 0x555555574440}, result = 0}, 
  struct_task__promise_type__FinalSuspend_0 = {struct_std____n4861__coroutine_handle_0 = {PointerType = 0x0}}, __coro_index = 0 '\000', struct_std____n4861__suspend_always_1 = {__int_8 = 0 '\000'}}
```

```
// tmp/t.cpp
#include <coroutine>

struct task{
  struct promise_type {
    task get_return_object() { return std::coroutine_handle<promise_type>::from_promise(*this); }
    std::suspend_always initial_suspend() { return {}; }
    std::suspend_always final_suspend() noexcept { return {}; }
    void return_void() noexcept {}
    void unhandled_exception() noexcept {}

    int count = 0;
  };

  void resume() noexcept {
    handle.resume();
  }

  task(std::coroutine_handle<promise_type> hdl) : handle(hdl) {}
  ~task() {
    if (handle)
      handle.destroy();
  }

  std::coroutine_handle<> handle;
};

 task coro_task(int v) {
  int a = v;
  __builtin_debugtrap();
  co_return;
}

int main() {
  auto k = coro_task(1);
  k.resume();
  return 0;
}
```

**Comment by Michael137** [comment]
Looks like this is a problem with how we are parsing artificial members out of DWARF (see [this FIXME](https://github.com/llvm/llvm-project/blob/main/lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp#L3062-L3067))
```
  // Don't turn artificial members such as vtable pointers into real FieldDecls
  // in our AST. Clang will re-create those articial members and they would
  // otherwise just overlap in the layout with the FieldDecls we add here.
  // This needs to be done after updating FieldInfo which keeps track of where
  // field start/end so we don't later try to fill the space of this
  // artificial member with (unnamed bitfield) padding.
  // FIXME: This check should verify that this is indeed an artificial member
  // we are supposed to ignore.
  if (attrs.is_artificial) {
    last_field_info.SetIsArtificial(true);
    return;
  }
```

When I speculatively remove those lines I get something slightly better (at least we see the child members properly now):
```
(lldb) v __coro_frame                                     
(_Z9coro_taski.coro_frame_ty) __coro_frame = {            
  __resume_fn = 0x0000000100003738                        
  __destroy_fn = 0x0000000100003970                       
  __promise = (count = 0)                                 
  v = 1                                                   
  __coro_index = '\0'                                     
  struct_std____1__suspend_always_0 = (__int_8 = '\0')    
  struct_std____1__suspend_always_1 = (__int_8 = '\0')    
}                                                         
```

Looks like time has come to do something about that FIXME

**Comment by vogelsgesang** [comment]
In case you are trying to debug some actual production code, it might be helpful to directly print the `__promise` variable (i.e. top-level `p __promise`; not the one nested inside `__coro_frame`). This value is literally your `task::promise_type` and will be printed properly.

### Slight tangent re. debugging of async coroutines

I assume your actual, non-simplified use case is to debug async coroutines since you named your coroutine `task` (and not `generator`)

For that use case, your `promise_type` usually contains a coroutine_handle which indicates where to resume execution after the current coroutine finished. You can hence get a pointer to your "async caller" by printing `__promise`. Inside this promise, you will see your `std::coroutine_handle` - for which lldb also has pretty-printing support. In fact, lldb also already devirtualizes type-erased `std::coroutine_handle`, so you should be able to essentially get a "async backtrace" if you use coroutines for async programming and you print your `__promise`.

See https://reviews.llvm.org/D132415 and https://reviews.llvm.org/D132624

When printing `std::coroutine_handle`, we currently also don't show you the internal state of your coroutine, but at least the contained promises should be displayed correctly.

Currently, the pretty printer devirtualizes a coroutine by looking at the `__promise` variable within the destroy-function of a coroutine (See https://github.com/llvm/llvm-project/blob/52db7e27458f774fa0c6c6a864ce197fa071a230/lldb/source/Plugins/Language/CPlusPlus/Coroutines.cpp#L69). I wanted to change this to use the `__coro_frame` type (as you are currently trying) because then I would have been able to also pretty-print internal state of suspended coroutines. However, I ran into a problem with that:

Apparently, g++ does not add the `__coro_frame` type to the destroy function, but only to the resume function. clang adds the `__coro_frame` type to both. However, we can't just use the resume function for devirtualization because at the final suspend point, the resume function will be set to a `nullptr`.

This could certainly be worked around (e.g., only doing devirtualization/pretty-printing based on `resume` as long as its available and then gracefully fall-back to using the `destroy` function if we are at the final suspend point). I didn't follow-through, though, since there were higher priority items internally in my team to drive the adoption of coroutines and because the basic pretty-printing of coroutines was good enough for the time being.

I hope this helps 🙂 
Unfortunately, I never found the time to time to update https://clang.llvm.org/docs/DebuggingCoroutines.html with this information

**Comment by hokein** [comment]
Thanks for all useful replies, they're helpful (sorry for being late, it is a bit strange that I don't get any email notifications).

> Looks like this is a problem with how we are parsing artificial members out of DWARF (see [this FIXME](https://github.com/llvm/llvm-project/blob/main/lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp#L3062-L3067))

Nice, thanks for finding it out. Is this issue different to resolve? If not, I'm happy to give it a try (though I don't know much about `lldb`). 

> For that use case, your promise_type usually contains a coroutine_handle which indicates where to resume execution after the current coroutine finished. You can hence get a pointer to your "async caller" by printing __promise. Inside this promise, you will see your std::coroutine_handle - for which lldb also has pretty-printing support. In fact, lldb also already devirtualizes type-erased std::coroutine_handle, so you should be able to essentially get a "async backtrace" if you use coroutines for async programming and you print your __promise.

The async stack trace is definitely the thing I'm looking for. Thanks for sharing the experience. Right, the `promise_type` in our internal coroutine library holds a caller `coroutine_handle`. As you mentioned, the async-stack-trace can be construct by walking up the chain.

While printing `__promise` is useful,  I think sometimes for debugging purposes, we need to inspect local information to understand the coroutine program state, having the ability to pretty print `__coro_frame` is helpful. For example, if a coroutine has multiple suspension points, to know which suspension point we are currently in, we need to inspect the `__coro_index` value in `__coro_frame`.

I'm looking at improving the debugging experience for an internal coroutine library. My initial idea is to write a LLDB python extension that prints the async-stack-trace and prints a particular coroutine frame (similar to the gdb script described in the llvm debugging coroutine page).

> See https://reviews.llvm.org/D132415 and https://reviews.llvm.org/D132624

I like the idea that printing the `continuation` recursively when printing the `__promise` (as described in `D132624`), but it was revert (to avoid a potential cycle issue). Now it just prints an address, then we need to use `p *(task::promise_type)(<promise_address>)` to see the internal content, which is a bit tedious.

> Currently, the pretty printer devirtualizes a coroutine by looking at the __promise variable within the destroy-function of a coroutine (See
> [llvm-project/lldb/source/Plugins/Language/CPlusPlus/Coroutines.cpp](https://github.com/llvm/llvm-project/blob/52db7e27458f774fa0c6c6a864ce197fa071a230/lldb/source/Plugins/Language/CPlusPlus/Coroutines.cpp#L69)
> ). I wanted to change this to use the __coro_frame type (as you are currently trying) because then I would have been able to also pretty-print internal state of suspended coroutines.

This looks like an improvement. I'm happy to take a look.

**Comment by Michael137** [comment]
> Nice, thanks for finding it out. Is this issue different to resolve? If not, I'm happy to give it a try (though I don't know much about lldb).

@hokein I haven't looked into fixing it yet. But patches are always welcome if you want to give it a shot!

I was thinking we will want to ask the language plugin about which artificial (i.e., compiler-generated) members we want to ignore. E.g., adding a `TypeSystemClang::ShouldIgnoreArtificialField` or something like that. Internally it would have a list of special names that shouldn't be skipped, like `__coro_frame`, etc.. (maybe there's a more robust heuristic?)

Then we can replace [this code](https://github.com/llvm/llvm-project/blob/a3c68754b089872ac6d5349cf24439a90835d196/lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp#L3062-L3067):
```
if (attrs.is_artificial) {
    last_field_info.SetIsArtificial(true);
    return;
  }
```
with a call to:
```
if (attrs.is_artificial && TypeSystemClang::ShouldIgnoreArtificialField(attrs.name)) {
    last_field_info.SetIsArtificial(true);
    return;
}
```

This would be my initial suggestion

**Comment by hokein** [comment]
@Michael137  thanks for the detailed guidance and pointers. I will take a look.

**Comment by vogelsgesang** [comment]
## Usage of the existing pretty printer

> > See https://reviews.llvm.org/D132415 and https://reviews.llvm.org/D132624
>
> I like the idea that printing the continuation recursively when printing the __promise (as described in D132624), but it was revert (to avoid a potential cycle issue). Now it just prints an address, then we need to use p *(task::promise_type)(<promise_address>) to see the internal content, which is a bit tedious.

**TLDR: Use `p --ptr-depth 9 __coro_frame`**

D132624 was not actually reverted (well, it was; but it was also reapplied). The latest comment in Phabricator which indicates that it was reverted again, was posted on the wrong review. Instead, D132815 and D132735 were actually reverted. 

The reason that `p __coro_frame` doesn't work out of the box, is that D132815 changed the pretty printer to no longer dereference the coroutine pointer implicitly. Instead, the pretty printer now returns a pointer-typed type. As all pointers, those can be expanded during printing by using the `--ptr-depth` parameter.

### Usage within an IDE

I assume that in the end your (internal) users of your internal library will be all using some IDE? Maybe Visual Studio Code? Maybe they are using [CodeLLDB](https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb)? If so, see below. If not, please describe your setup and still see below - maybe it still applies to your IDE, too...

In general, the "variables view" should nicely display you the async call stack.
You need to add the `__coro_frame` variable to your "watched variables", though. It doesn't show up in the variables view by default because the `__coro_frame` variable - surprise! - is marked as artificial and hence ignored by LLDB by default.
You can uncollapse your "continuation" `std::coroutine_handle` from the variable tree as far as you want.

## Duality between `ShouldIgnoreArtificialField` and `ShouldIgnoreArtificialVariable`

As mentioned in the previous paragraph, `__coro_frame` is hidden by default, e.g., from the output of `frame variable`. This makes it quite hard to discover the presence of this variable. I think we should reveal this variable by default, to make it more easily discoverable. I gues, this would require the introduction of `ShouldIgnoreArtificialVariable` or something similar...

@Michael137 Given that we probably want both `ShouldIgnoreArtificial{Field,Variable}`, does that change your recommendation to put `ShouldIgnoreArtificialField` into `TypeSystemClang`? Not sure where exactly LLDB is ignoring artificial stack variables... Is it fine to access a callback from `TypeSystemClang` from whereever that place may be?

**Comment by Michael137** [comment]
> @Michael137 Given that we probably want both ShouldIgnoreArtificial{Field,Variable}, does that change your recommendation to put ShouldIgnoreArtificialField into TypeSystemClang? Not sure where exactly LLDB is ignoring artificial stack variables... Is it fine to access a callback from TypeSystemClang from wherever that place may be?

The logic to keep the artificial variable out of the frame is [here](https://github.com/llvm/llvm-project/blob/072a7edec3ca3729ec597c413d2fcc9557d88ffc/lldb/source/Commands/CommandObjectFrame.cpp#L686-L689).

Inside that check there we query the language plugin already:
```
if (auto *runtime: LanguageRuntime * = process->GetLanguageRuntime(GetVariable()->GetLanguage()))
  if (runtime->IsAllowedRuntimeValue(GetName()))                                                     
    return false;                                                                                          
```

For C++, `IsAllowedRuntimeValue` is currently implemented as follows (see [here](https://github.com/llvm/llvm-project/blob/072a7edec3ca3729ec597c413d2fcc9557d88ffc/lldb/source/Plugins/LanguageRuntime/CPlusPlus/CPPLanguageRuntime.cpp#L43C38-L45)):
```
bool CPPLanguageRuntime::IsAllowedRuntimeValue(ConstString name) { 
  return name == g_this;                                           
}                                                                  
```

So we can introduce a single API in `TypeSystemClang` that checks for the compiler generated names we shouldn't ignore. Then we can:
1. use it in `DWARFASTParserClang` can query to make sure we populate the fields correctly
1. use it in `IsAllowedRuntimeValue` (which will make `frame var` show artificial vars)

**Comment by hokein** [comment]
Thanks for the insights, everyone.

> TLDR: Use p --ptr-depth 9 __promise

Glad to know there is an easy way to print the `__promise` recursively (I were not aware of `--ptr-depth`).

I think it should be `frame variable --ptr-depth 9 __promise`, `--ptr-depth` doesn't work with the `print` command.

> Usage within an IDE
> I assume that in the end your (internal) users of your internal library will be all using some IDE? Maybe Visual Studio Code? Maybe they are using [CodeLLDB](https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb)? If so, see below. If not, please describe your setup and still see below - maybe it still applies to your IDE, too...

> In general, the "variables view" should nicely display you the async call stack.
> You need to add the __promise variable to your "watched variables", though. It doesn't show up in the variables view by > default because the __promise variable - surprise! - is marked as artificial and hence ignored by LLDB by default.
> You can uncollapse your "continuation" std::coroutine_handle from the variable tree as far as you want.

Good point. We use a similar variant lldb vscode extension internally. Having the `__promise` variable in the variables view by default is definitely an improvement.

> So we can introduce a single API in TypeSystemClang that checks for the compiler generated names we shouldn't ignore. Then we can:
> 1. use it in DWARFASTParserClang can query to make sure we populate the fields correctly
> 2. extend IsAllowedRuntimeValue (which will make frame var show artificial vars)

This makes sense.

For 1, fields in a coroutine frame are varied (local variables, compiler generated members etc), and we want to print all of them. It seems not easy to have a completed whitelist. One solution is that we populate all fields if the `class_clang_type` is a coroutine frame type (we could use a simple heuristic -- the name ends with `.coro_frame_ty`).

**Comment by Michael137** [comment]
> For 1, fields in a coroutine frame are varied (local variables, compiler generated members etc), and we want to print all of them. It seems not easy to have a completed whitelist. One solution is that we populate all fields if the class_clang_type is a coroutine frame type (we could use a simple heuristic -- the name ends with .coro_frame_ty).

Good point, I was originally thinking we'd allow any artificial names that aren't on a "ignore_list". E.g., we know that we definitely don't want to consider artificial variables with `_vptr$` in the name. So maybe that's a good enough heuristic? But hard to know if we're missing something from that list. Your suggestion of checking the `class_clang_type` seems reasonable (maybe adding a new `CPPLanguageRuntime::IsAllowedRuntimeValue(CompilerType)` overload. Although I'm not sure we have such `CompilerType` available in `CommandObjectFrameVariable` (?)

**Comment by hokein** [comment]
I made a draft patch at https://github.com/llvm/llvm-project/pull/70779. It doesn't have unit tests yet, let me know what you think.
